### ISSUE TITLE
Client certificate verification callback

### ISSUE THREAD

**Issue by michal-tomlein** [apple/swift-nio-ssl#25]
Hi,

Are there any plans to support authentication using client certificates?

Ideally, I’d like to be able to provide a callback for this that will be called with the client’s certificate and public key (I’m specifically interested in the public key). This could be a new `CertificateVerification` enum case that I could pass to `TLSConfiguration.forServer(…)` instead of `.none` (the current default).

If the callback succeeds, I’d also need a way to pass the certificate and public key to other handlers in the pipeline. I’m new to SwiftNIO, so I’m not sure what the right way to do this would be. Perhaps this association of the channel with the certificate/public key could be established directly in the callback.

Thanks.

**Comment by Lukasa** [comment]
@michal-tomlein I certainly think we should add support for this, yes, but I'm a bit uncertain about what specifically you want to achieve. Can you explain your use-case, and why simply adding an appropriate trust root doesn't do the job?

**Comment by michal-tomlein** [comment]
Thanks for the quick response. We’re using self-signed client certificates for device authentication. Each device connecting to our TCP server authenticates using its client certificate. The public key in the certificate is matched against a database of preapproved public keys.

The keys are generated by and stored in keychain using `SecKeyCreateRandomKey`. The private keys never leave the client devices, but the public keys are shared with the server. The keys are wrapped in self-signed certificates and used for client authentication.

**Comment by Lukasa** [comment]
Aha, I see, the certificate is just basically a mechanism by which you transmit proof that the remote peer controls a known private key.

Yeah, we can add this support. You'll need quite a few more APIs than we currently provide though, as right now we don't provide any access to *details* about a certificate: `OpenSSLCertificate` is basically just an opaque handle to an OpenSSL structure.

**Comment by michal-tomlein** [comment]
Thanks, that would be great!

**Comment by michal-tomlein** [comment]
I’m fast approaching a point when I’ll need this. What’s your estimate for when this could be implemented? I don’t mean to be impatient, just trying to decide whether I should find an alternative approach in the meantime. Thanks!

**Comment by Lukasa** [comment]
We can probably add it sometime in the next couple of weeks: what’s your schedule look like?

**Comment by michal-tomlein** [comment]
I’ll be working on this starting Monday. If you can add it sometime next week or even the week after that, I would definitely choose not to ship a workaround.

**Comment by Lukasa** [comment]
What details do you need to access about the certificate in the callback?

**Comment by michal-tomlein** [comment]
Just the public key bytes for our use case. I guess it would be nice to expose more details over time to have a more general purpose interface, but for us it’s only about the public key.

**Comment by Lukasa** [comment]
Ok, great answer, that's definitely not too much work. I'm reluctant to go too far down the road of exposing certificate data here because it starts pushing swift-nio-ssl into the territory of a general-purpose crypto library, but exposing public key bytes is a bit more constrained.

When you say "public key bytes", do you want the entire Subject Public Key Info block?

**Comment by michal-tomlein** [comment]
OK, sounds good.

**Comment by Lukasa** [comment]
See #49.

**Comment by michal-tomlein** [comment]
Fantastic! Thanks. I’ve tested it and it works great.

Have you considered adding an enum case such as `.callbackVerification` to `CertificateVerification` that would function similar to `.none` (no validation against the trust store), but would still use the callback?

**Comment by Lukasa** [comment]
I have, but I rejected it for a number of reasons.

The first is that it's a source breaking change: adding enum cases is always source-breaking in Swift, and so requires a Semver Major release. That's a reason to investigate alternative approaches.

The second reason is that it makes doing anything with the certificate hard. As it's set globally on the `TLSConfiguration` object, which is intended to be shared across many different connections, we'd have to give you some per-connection state in order to work out what `Channel` a given callback applies to. The only thing we could give you that would be *useful* is the `Channel` object itself, and even there it's not easy for you to get to the things you actually want from there.

The advantage of setting this callback in the handler is that you can close over whatever state your callback needs. This means I don't have to try to guess what per-connection state you want: you can just ensure you have it.

This is such a substantially better user experience in the closure that it seemed to be a killer reason to do it this way.

**Comment by michal-tomlein** [comment]
Thanks for the explanation. I did not realise it’s a source breaking change.

Regarding the second reason, I should have been clearer: having tried your implementation, I totally agree you made the right call adding the callback to the handler for all the reasons you mention.

I did not mean to imply that the callback should be an associated value of the new enum case. I was just looking for a way to avoid any validation against the trust store, but still have the callback invoked. I am not sure what the performance impact of such validation is. Perhaps it’s negligible, but it is still unnecessary work for a server handler. As it stands, the callback is invoked with a result value of `.failed`.

An alternative would be to have the callback invoked even if `CertificateVerification.none` is specified.

**Comment by michal-tomlein** [comment]
I see, thanks. The callback seems to be invoked twice, but I’m assuming this is part of normal operation. When do you expect this to be merged into master?

**Comment by Lukasa** [comment]
As the apidoc notes, the callback will be invoked once per certificate in the chain, so you shouldn’t expect there to be only one invocation in most cases.

We’re waiting on code review from @weissi, but probably sometime this week.

**Comment by Lukasa** [comment]
Merged this to master, we can close.

**Comment by michal-tomlein** [comment]
Thank you for your work on this. This is likely a fairly niche feature, but a very important one for those few who need it.

**Comment by wafflespeanut** [comment]
Hi @Lukasa! Thanks a lot for implementing this feature!

I have a question on this:

> Unfortunately, OpenSSL will not invoke the verify callback at all if certificate validation is disabled. That’s because for OpenSSL to invoke the callback according to its documentation it needs to first build a trust chain to validate.

If OpenSSL always requires a bunch of trust roots for verifying the certs, does it mean that it's impossible for a client to present a self-signed cert?

**Comment by Lukasa** [comment]
No. Clients can present self-signed certs if the self-signed cert is itself in the trust root, or if the verification callback manually validates that self-signed certificate, or if certificate verification is disabled.

**Comment by wafflespeanut** [comment]
Okay, my bad, I should've seen the PR diff. That's exactly what I wanted - to be able to manually validate the certificate. Thanks!

**Comment by Sn0wfreezeDev** [comment]
Hey, 
I see that pull request has been merged and I tried to integrate the custom verification callback, but I cannot manage to have it called. 
When I make a request using my custom client certificate if fails during the verification. As far as I understand the callback it will still perform the normal handshake and call it afterwards. So even if the verification fails the callback should be called in my understanding. 
As I looked through the nio-ssl implementation I can also confirm that all functions that reference the callback are never executed. 
Is this feature still supported or has it become deprecated after the upgrade to boringssl? 

Here's my source code: 

```swift
                    if var tlsConfiguration = tlsConfiguration {
                        tlsConfiguration.applicationProtocols.append("http/1.1")
                        let sslContext: NIOSSLContext
                        let tlsHandler: NIOSSLServerHandler
                        do {
                            
                            sslContext = try NIOSSLContext(configuration: tlsConfiguration)
                            tlsHandler = try NIOSSLServerHandler(context: sslContext, verificationCallback: verificationBlock)
                        } catch {
                               ...
                        }
                        
                        return channel.pipeline.addHandler(tlsHandler). .... 
```

The TLS config: 
```
        
        let tlsConfig = TLSConfiguration.forServer(
            certificateChain: [.certificate(certificate)],
            privateKey: .privateKey(key),
            maximumTLSVersion: .tlsv12,
            certificateVerification: .noHostnameVerification,
            trustRoots: NIOSSLTrustRoots.default
        )
```

The callback is set in this line: 
https://github.com/apple/swift-nio-ssl/blob/f0b118d9af6c4e78bc4f3f4fbb464020172b0bf4/Sources/NIOSSL/SSLConnection.swift#L132 

This will set the verification callback in the C sources here: 
https://github.com/apple/swift-nio-ssl/blob/f0b118d9af6c4e78bc4f3f4fbb464020172b0bf4/Sources/CNIOBoringSSL/ssl/ssl_x509.cc#L649 

During my debugging this callback in C is actually never called. 
It should  be called here: 
https://github.com/apple/swift-nio-ssl/blob/f0b118d9af6c4e78bc4f3f4fbb464020172b0bf4/Sources/CNIOBoringSSL/ssl/ssl_x509.cc#L362 

But this method is never executed. 

Kind regards 
Alex

**Comment by Lukasa** [comment]
What platform are you running on @Sn0wfreezeDev? This misbehaviour is definitely a bug, but we should be able to mitigate it by using the new callback API. Of course, the real mistake here is that we never exposed the new callback API for servers 🤦 . I'll throw together a patch.

**Comment by Sn0wfreezeDev** [comment]
Thank you for the quick reply @Lukasa. I am using macOS and I also want to run this code on iOS later. I am trying to use swift-nio-transport services now for my server code. This could solve the issue for now.

**Comment by Lukasa** [comment]
Yes, we recommend using swift-nio-transport-services if possible: you'll have a much better all-around experience. However, we should definitely fix the two issues you've outlined, so thank you for bringing those to our attention!

**Comment by Lukasa** [comment]
For the first issue, see #226. This will provide the supported-going-forward solution to the issue. Next I'm going to investigate why the existing callback isn't being called, as it ought to be in your case.

**Comment by Lukasa** [comment]
Ok, I don't reproduce the second issue. When I write my test, the verification callback is definitely called. Is the client presenting certificates at all? When we turn on certificate verification, we require that the client actually present some certificates (we set `VERIFY_FAIL_IF_NO_PEER_CERT`) which could well be interfering with the logic here.

**Comment by Sn0wfreezeDev** [comment]
Looks good! That's a fast development lifecycle 👍

**Comment by Sn0wfreezeDev** [comment]
@Lukasa  you are right. The client is sending its certificate only if the server specifies that peer certificates are required. Is this option available in nio-ssl?

**Comment by Lukasa** [comment]
nio-ssl sets that option unconditionally if you turn on certificate verification. You can see this if you take a pcap of the exchange. When I enable certificate verification in the example `NIOTLSServer` and then attempt to connect it, Wireshark shows the server sending a `Certificate Request` record, which is the request for the client to provide a certificate:

<img width="738" alt="Screenshot 2020-06-16 at 16 59 10" src="https://user-images.githubusercontent.com/1382556/84798312-b4e3dd00-aff2-11ea-8ba9-297cde68f8fc.png">
